File: haskell/src/Solutions/Euler045.hs
module Solutions.Euler045 where
import Utils.List(ascendingIntersection)

tri = map (\x -> x * (x + 1) `div` 2) [286..]
pen = map (\x -> x * (3 * x - 1) `div` 2) [166..]
hex = map (\x -> x * (2 * x - 1)) [144..]

euler045 :: IO String
euler045 = return $ show $ head $ tri `ascendingIntersection` pen `ascendingIntersection` hex

File: rust/src/solutions/euler045.rs
use crate::utils::numeric::sequences::{hexagonal, pentagonal, triangular};

pub fn euler045() -> String {
    let mut tri = triangular::Triangular::<u64>::new().skip(285).peekable();
    let mut pen = pentagonal::Pentagonal::<u64>::new().skip(165).peekable();
    let mut hex = hexagonal::Hexagonal::<u64>::new().skip(143).peekable();
    loop {
        let t = tri.peek().unwrap();
        let p = pen.peek().unwrap();
        let h = hex.peek().unwrap();
        if t == p && t == h {
            break t.to_string();
        } else if t <= p && t <= h {
            tri.next();
        } else if p <= t && p <= h {
            pen.next();
        } else {
            hex.next();
        }
    }
}
